notebooks/formatnb.py

```diff
@@ -0,0 +1,51 @@
+import nbformat
+import black
+
+
+def clean_outputs(nb):
+    for cell in nb.cells:
+        if cell["cell_type"] == "code":
+            cell["execution_count"] = None
+            cell["outputs"] = []
+    return nb
+
+
+def black_code(nb):
+    fm = black.FileMode()
+    for cell in nb.cells:
+        if cell["cell_type"] == "code":
+            try:
+                cell["source"] = black.format_str(cell["source"], mode=fm)
+            except black.InvalidInput:
+                continue
```

Comment:
Perhaps print something about possibly invalid python code encountered?

Reply:
The %cell_magic lines and !shell_execution lines trigger that exception...

Reply:
Something like this: 
```
+++ b/notebooks/formatnb.py
@@ -1,5 +1,6 @@
 import nbformat
 import black
+from sys import exit


 def clean_outputs(nb):
@@ -12,12 +13,19 @@ def clean_outputs(nb):

 def black_code(nb):
     fm = black.FileMode()
-    for cell in nb.cells:
+    for cellnum, cell in enumerate(nb.cells):
         if cell["cell_type"] == "code":
             try:
                 cell["source"] = black.format_str(cell["source"], mode=fm)
             except black.InvalidInput:
-                continue
+                print(f"Possible invalid python encountered in cell {cellnum}")
+                print(f"cell source causing exception:\n{cell['source']}\n")
+                print(r"Script will abort (leaving your file unchanged) unless you type 'continue'")
+                abort = input()
+                if abort == 'continue':
+                    continue
+                else:
+                    exit()
     return nb
```

Reply:
Notebook magics are not valid Python. Most cell magics can be isolated to their own cell.

I also don't think we want to make the tool interactive. It shouldn't add cognitive overhead to the user.